Hail as a service, first steps

akotlar opened the pull request:
## Basic philosophy
https://developers.google.com/web/fundamentals/performance/prpl-pattern/

Hybrid server-side client-side rendered application, with eager pre-loading of resources. First page visited is server rendered. All client-side code asynchronously fetched, using service workers if available. HTML is "hydrated"/bound by React, and from then on has the responsiveness of a client-side application.

This is what we have in the current pull request. Initial view / first page ready in ~10ms, DOMContentLoaded in ~60-120ms (excluding network latency).

## Why not static/HTML web?
In practice: there is no such thing. Even document-centric sites often need dynamic templates, and will therefore use PHP, Python, NodeJS, Go, Rust, etc. These only work on a server, and only serve interpolated, static documents. Any interactive elements require Javascript.

As soon as you need Javascript, the choice becomes Vanilla JS, JQuery, or something more structured. Vanilla JS requires a lot of boilerplate (verbose event binding, DOM modification, needs polyfills since browser incompatibilities). JQuery makes this easier, but is 1) very slow, 2) provides no structure. Vanilla JS and JQuery tend to devolve to soup of global state-modifying code, with a lot of time spent on figuring out how to update values in DOM elements. 

React/Next make DOM modification declarative, and very very easy. They provide a great deal of structure (especially with Next handling tooling), and thanks to the virtual dom / reconciliation process, performs, in many cases, much faster than directly modifying the DOM (HTML) (i.e plain JS). React also handles necessities like properly escaping all inputs, for XSS attack prevention. All of this in a bundle size that isn't significantly bigger than JQuery, without all of those benefits (and React is rapidly shrinking).

It's possible to avoid Javascript. One can simulate interactivity by issuing a server GET request for a new page, i.e click on a link with a GET variable ?someVar=val and get a new page. This is slow (full round trip cost), and puts much more load on the server (since it not only needs to make the db call, but interpret PHP/Python to render the view). 

There is a good reason why JS and monolithic single page applications became popular, with all of the initial-load (bundle size) downsides: client-side rendering allows perceived performance on the order of native mobile or desktop applications. Achieving interactive UI's without JS or Web Assembly, by using server-rendered pages, is ~impossible.

We will achieve this interactivity without suffering the bundle-size-before-first-render cost, at minor developer costs vs server-side-only rendering.

Lastly, it is possible to abuse any technology. Javascript brings to mind "bloated"; this is an implementation issue. PHP/Python/Perl websites also used to be slow and ugly (Geocities).
  * NodeJS/Javascript/V8 JIT is consistently faster than PHP, Python, and ~Java:  https://www.techempower.com/benchmarks/

## Why NodeJS, React, etc
1. Javascript is the only language supported by modern browsers. Web assembly will change this (compile target == web assembly, language == rust | go | python), but is not nearly as mature
2. Ecosystem. Chosen technologies are (likely) by far the most popular. We should quantify this better
3. Performance. NodeJS is faster than Flask, React is ~fastest JS view layer. Next makes it really easy to split app into page bundles, and (on localhost) achieves DOMContentLoaded of ~70-100ms, and faster interactivity: first loaded page (the page of the current route) is ~6-10ms.
    * [Techempower]: https://www.techempower.com/benchmarks/
    * [Node vs , ](https://medium.com/@mihaigeorge.c/web-rest-api-benchmark-on-a-real-life-application-ebb743a5d7a3)

    * React vs other client side micro bench (pay attention to "Non-keyed"): https://krausest.github.io/js-framework-benchmark/current.html
4. Structure, aforementioned
5. Path to relatively performant desktop and mobile applications, via [Electron](https://getstream.io/blog/takeaways-on-building-a-react-based-app-with-electron/). [Visual Studio Code](https://github.com/Microsoft/vscode) and [Slack](https://slack.engineering/growing-pains-migrating-slacks-desktop-app-to-browserview-2759690d9c7b) are good examples. Facebook Messenger written in React Native, which we have an even more straightforward path to.
6. Low cognitive cost (relative to Angular, others. React is just a view layer, and has a tiny API.

I've develop a large application in AngularJS, and have spent a bit of time with Angular2+. There is no comparison: Angular takes months to know well, React days at worst.

Also, by not buying into a full framework, we achieve modularity: If we end up finding React too slow, even with [planned 2019 improvements](https://reactjs.org/blog/2018/11/27/react-16-roadmap.html), there are plenty of others view layers we can migrate to, without gutting our entire app.

Next makes this somewhat trickier, but since it's mostly a light wrapper around Webpack, there is essentially no Next-specific code (just a bit in pages/_app.js and pages/_document.js). Migrating from Next won't be an issue, we would just lose some of the tooling benefits.

8. Typescript: static typing, syntax more familiar to OO-language devs (interfaces, types, classes).

9. Really amazing debug, tooling. React debug tools natively included in Chrome for instance. First class support for JSX (React-wrapped HTML) in popular IDE's, most obviously Visual Studio Code.

## Tech TL;DR
Mainly React.

React should take about a day to learn well enough to make contributions.

Guide: https://reactjs.org/docs/introducing-jsx.html

```jsx
# Renders Hello World
# Biggest annoyance (may go away in 2019) is that "class" is not a valid tag (reserved by React)
export default function SomePage() {
 const name = 'Alex'

 # Renders "Hello Alex"
  return (
    <div className='some-class'> Hello {name} </div>
  )
}
```

## Challenges
1. Auth 
Authentication is tricky, but not for any reason specific to React, Next, Node. Server-side rendered apps tie the web app to the resource server; as such it's easier to hide sensitive information. 

Mobile and desktop apps have dealt with this for 2 decades. We should build a robust infrastructure, and not one that requires server-rendered web pages for security.

Currently it seems Auth0 may not be the best choice: it does not interface for us with third-party API's; requires us to either insecurely store 3rd party access tokens (with at least 1 extra round trip), or altogether proxy all third-party requests through our own resource server... Firebase Auth seems to avoid these limitations.

## TODO:
1. Create a structured description of this pull request
2. Incorporate Firebase Auth in place of Auth0 for 3rd party access token benefits.
3: Scorecard
  3a. Draft working GraphQL V4 scorecard implementation
  3b. Finish authenticated GraphQL V4 scorecard implementation
4. Batch
  4a: Setup dev batch endpoint
  4b: Call batch endpoint (no auth), and return any data
  4c: List all available jobs
     * By querying Batch api, or Kubernetes directly
  4d: Receive current status of 1 job
  4e: Authentication
  4f: Polish (longest step): make interacting with batch achievable within perceived 16ms.
      * goal: subscribe to events in web socket
      * may want to save user job state in a Hail-controlled database (possible to use Firebase or Mongo, may prefer relational db, maybe Postgres or MySQL).
  4other:  Figure out state question (sufficient to use Kubernetes)
5. Basic notebook interface.
6. Connect websocket logic (non-GraphQL)
7. Authenticate web socket via Oauth2
8. Incorporate GraphQL subscriptions (first: GitHub API)
9. Write tests
10. Mock GraphQL endpoints
11. Integrate web and api server bits into CI (maybe should be prioritized earlier...I prefer to get draft of major functions done first; am new to writing tests for React/Node).

## Near-term goals (<= 6 mo)
1. Upload, download
2. Launch clusters, pay for them
3. ?

## Longer-term goals
1. Much simpler interface to Hail. I would like steps that can be performed without programming to be done so. I would prefer fasta->variant filtering to be done as in Bystro (at least from the interface standpoint), i.e without opening up a notebook. Common analyses pipelines should also be possible without any interaction with a python notebook: GWAS, rare-variant (SKAT) analyses have, it seems, relatively few permutations. Those should be behind UI primitives. At each stage of a 
2. Social network bits: users should be able to share job state with other users (requested by Bystro users on 22q consortium project) at the least.
3. Record job state using something like Merkle tree. Checkout state. Aka "blockchain"
4. Cooperative analysis: provide system for people to validate analyses
 
Basic idea: 

1) People donate computational resources for ad-hoc heterogenous clusters. 
2) People donate intellectual capital. Re-run analyses without the full available code. See if they can replicate (not p-values, but order). Could generate multiple-hypothesis-test corrected aggregate. These users get publication credit as consortia
3) People donate minor intellectual capital: Re-run analysis with full available code. Report on success. This will catch bugs, and non-deterministic results (for instance, if reported accuracy depends on local minima..similar or better minima may only occur once in a great while). Similar to 2.


## Timetables
1-3a: 12/10/18
3b: by 12/15/18
4a-4b: by 12/12/18
4c-d: by 12/15/18


This probably shouldn't be merged for a while. Still working on authentication handling for third party APIs. All first party APIs (our stuff) is well controlled, can be extended from existing codebase.

akotlar commented:
Example of tooling/debug (Firefox, built in to Chrome as well)

<img width="1308" alt="screen shot 2018-12-11 at 9 57 24 pm" src="https://user-images.githubusercontent.com/5543229/49844151-c0b46a80-fd8f-11e8-8ed5-de0d8293d538.png">

Hidden cost of isomorphic SSR in React: small bundle hit: currently sends serialized state with initial / first page render. In case of scorecard, costs a few KB. No such cost incurred from client-side requests, but if the page that fetches those resources is the first-hit page, would incur loading indicator (on any other page could be prefetched).

danking commented:
should the `haas/packages` folder be checked in?

akotlar commented:
> should the `haas/packages` folder be checked in?

I don't quite understand. haas is a monorepo containing all the bits that will make up the web/mobile/desktop/auth api and any other services that allow users to interact with Hail without installing it themselves. 

https://github.com/hanford/next-offline#service-worker-path

Everything in packages is a separate package (web, api1), separately publishable to npm, but which allows dependencies to be shared.

Does that help?

akotlar commented:
Keeping in mind Cotton's queries last week, researched and found much lighter alternative to ExprsesJS for the server api. A few years ago, Express had low impact on node performance; it has become bloated.

Found a light (~200 LOC) "framework" called Polka, that is small enough to maintain ourselves. It mainly adds light route-matching capabilities, to avoid repeating boilerplate when writing the Node server. Easy to follow. It's also the fastest "framework" available, outside of C/Go/Rust. Matches Falcon, and allows 1 language for server/web. (Also Node has a far larger ecosystem).
  * https://github.com/the-benchmarker/web-frameworks 
  * Polka also nearly compatible with Express's middleware api, so many existing packages are either directly usable, or with minor modifications. This was a desire of mine, since nearly everything server-y for node is really written for Express.

Last commit removes all Express, adds a rewritten express-jwt for access token verification, and shows client credential exchange, backed by Redis cache, for <=4ms fetching of

akotlar commented on haas/packages/hail-api-gateway/api/github/index.js:
```diff
@@ -0,0 +1,74 @@
+    if (!(ctx && ctx.graphqlContext && ctx.graphqlContext.user)) {
+      return {
+        headers: {
+          Authorization: `bearer ${tokenForScheamIntrospection}`
```
typo, fix

akotlar commented on haas/packages/hail-api-gateway/api/user/lib/auth/lib/middleware.js:
```diff
@@ -11,7 +11,7 @@ const InvalidTokenError = require.main.require(
 
 const redisClient = require('redis').createClient({
-  parser: 'hiredis'
+  // parser: 'hiredis'
```
by recommendation of the node-redis client

chrisvittal commented on haas/packages/public/.vscode/settings.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "python.pythonPath": "/usr/local/opt/python/bin/python3.7",
```
This seems like it shouldn't be committed to the tree. I for one, do not have python at this location.

akotlar replied:
Yep, good catch!

akotlar commented:
Currently pruning dependencies, forking NextJS to remove poly fills for older browsers.

Investigated using Inferno.js as a lighter alternative to React. Saves ~20-30KB bundle size, and is somewhat faster. However, main Inferno dev moved to React core team, and React is focusing on the optimizations present in Inferno for 2019 (DOM: move to native events where possible), as well as introducing optimizations not found in Inferno (compile time targets: initially inlining, future maybe web assembly binaries; move rendering work to separate thread / concurrent rendering). Furthermore, React ecosystem is orders of magnitude larger, so we save a huge amount of dev time by avoiding Inferno, and have greater likelihood of LTS.

Notably, I realized that most of my bundle size was coming from inefficient bundling of Material UI and due to Apollo's insanely large graphQL bundle. Removing these now.

Lastly, React is actually very efficient. jQuery is ~31.1KB minified. React is 3KB, while React DOM is 33.8KB. In 2019, React DOM will shrink. In any case, given that React is both faster than jQuery, dramatically simplifies development, and introduces development structure, 4KB cost is imo worth it.

akotlar commented:
Bundling working well now:

For instance, the addition of this file, which handles the auth0 callback/sets cookie, adds only *501B* despite importing Auth and react-easy-state :tada:

```jsx
// TODO: Replace Loading component without Material UI
import { Component } from 'react';
import Router from 'next/router';
import { view } from 'react-easy-state';
import Auth from '../lib/Auth';

class Callback extends Component {
  componentDidMount() {
    Auth.handleAuthenticationAsync(err => {
      // TODO: notify in modal if error
      if (err) {
        console.error('ERROR in callback!', err);
      }

      Router.push('/');
    });
  }

  render() {
    return !Auth.isAuthenticated() ? <div>Loading</div> : <div>Hello</div>;
  }
}

export default view(Callback);
```

<img width="353" alt="screen shot 2018-12-19 at 5 06 59 pm" src="https://user-images.githubusercontent.com/5543229/50251076-ad695680-03b0-11e9-88f2-28d3ff7daa33.png">

akotlar commented:
Further pruned. Removed all GraphQL libraries, besides graphql-tag, which I like, because 1) simple hash-based cache: no need to walk complex graph to normalize cache, because in most cases I'm perfectly fine with not re-using cache across different queries (that may have some shared fields). Apollo does something "smarter", but much slower: walks a query, checks that the requested fields for a node are the same, and that the node's id is the same, as some other query. 

Also removed react-icons... I was going to use this in place of material-icons, because I thought loading the full font, when I needed only a few icons, would be unnecessarily expensive. It turns out that I cannot find a library where a single icon import (react-icons or MaterialUI) is smaller than Google's entire material design font: a single font (there are several needed to cover all icons) is ~500B. A single react-icons icon is ~2KB on dev (production may be smaller due to tree shaking). Also, am opposed to CSS-in-JS: slower, worse tooling, larger. Benefits are dynamic selectors, which are really no advantage that I can see (without them can still dynamically apply classes, as in the yee ol days of pleb vanilla js).

<img width="1445" alt="screen shot 2018-12-19 at 11 37 01 pm" src="https://user-images.githubusercontent.com/5543229/50264368-3fd91c80-03e8-11e9-8c73-d3ff4744ff85.png">

akotlar commented:
Research on UX/UI, and impact on customer acceptance

1) http://credibility.stanford.edu/pdf/p80-fogg.pdf

akotlar commented:
Oh, and the app is meant to operate behind HTTPS; when deployed, running the web app with 
`npm run start` instead of `npm run prod-test` will enable secureOnly cookies.

akotlar commented:
@danking Set auth0 callback based on window (this turns out to be cleanest), cleaned up styles, further reduced bundle size a bit by removing use of a state management tool (that wraps any arbitrary object in an observable that can be watched), and added a basic header menu to allow logout.

I think the most challenging part of using this web architecture will be managing actions on server vs browser. That is probably the only piece that isn't obvious.

akotlar commented:
Also, interesting to note: 

Home page (with menu bar, dark icon, not logged in): 1.5KB .gz . Logged in: 3.1KB

Bundle size: on order of 100KB. However, 30% of this is the auth0 client library; we can modify it to save space. I've commented on an issue with some light guidance on how to save 5.5KB of that.

Effectively 70KB for React + React-Dom + Webpack tooling + all page js compares quite favorably with a jquery-only solution, while being [~2x faster](https://github.com/jonmiles/react-performance-tests), far easier to manage, and with a much large ecosystem (and jquery-only solution would do nothing for universal rendering). The React side should drop this year substantially. They are also interested in writing a compiler to completely remove the vdom, compiling to optimized javascript or maybe web assembly. That may be something interesting to us as well.

akotlar commented:
Regarding SSR-only mode. This is the default behavior.

SSR is mostly a function of routing. If we allow the client to handle routes, we save the roundtrip in reconciling current app state (current DOM) with the next state (next page's DOM).

To "enable" this functionality, instead of using <Link> use <a>

Nextjs has excellent documentation and a responsive maintainer base: https://github.com/zeit/next.js/issues/575
